packages/babel-register/src/node.js

```diff
@@ -6,7 +6,73 @@ import * as babel from "@babel/core";
 import { OptionManager, DEFAULT_EXTENSIONS } from "@babel/core";
 import { addHook } from "pirates";
 import fs from "fs";
-import path from "path";
+import path, { extname } from "path";
+
+// Technically we could use the "semver" package here, but (for exmaple)
+// parseFloat("4.23.6") returns 4.23 so it's "good enough"
+const BABEL_SUPPORTS_EXTENSIONS_OPTION = parseFloat(babel.version) >= 7.12;
+
+// The "pirates" library, that we use to register require hooks, does
+// not allow hooking into all the loaded files, but requries us to
+// specify the extensions upfront.
+// In order to hook into all of them, we need to do two things:
+//  1) Add hooks for all the already registered extensions, defined in
+//     Module._extensions. By doing so, we can shadow the loaders already
+//     defined.
+//  2) Node throws an error when requireing .mjs unless a hook has been
+//     defined. For compatibility reason (@babel/register can load .mjs files)
+//     we add the extension to the list.
+//  3) Node fallbacks to the .js loader for unknown extensions, however
+//     pirates will only run our hook if it actually matches the extension
+//     (without a fallback mechanism). However, it checks if an extension
+//     has been registered by checking 'extensions.indexOf(...)'.
+//     And... we can make it always return true! ^-^
+// Since this is not technically part of the public API of "pirates", the
+// version in package.json is fixed to avoid untested updates.
+function generateExtensionsArray(exts) {
+  return Object.defineProperty(Array.from(exts), "indexOf", {
+    configurable: true,
+    writable: true,
+    enumerable: false,
+    value: () => true,
+  });
+}
+
+// Node.js's algorithm tries to automatically resolve the extension of the
+// required file. This means that if you have, for example, require("./foo")
+// it can load foo.js (even if ".js" is not specified).
+// This doesn't only work with predefined extensions: whenever a new extension
+// hook is registered, Node.js can resolve it.
+// For this reason, we need to register new extensions as soon as we know about
+// them. This means:
+//   1) When we compile a file and see a new extension in its `extensions` option.
+//   2) When a file with a new explicit extension is loaded
+//
+// In practice, this means that to load foo.ts you have to use
+//    node -r @babel/register ./foo.ts
+// instead of just
+//    node -r @babel/register ./foo
```

Comment:
This will need to be clearly written in the docs.

Another example that cannot work:
```js
// index.js
require("@babel/register");
require("./foo"); // <-- You'll need require("./foo.ts") here, since it's the first time Babel sees a .ts file

// foo.ts
console.log("OK");
```

Note that the ESM loader won't have this problem, it's CJS-specific (and only for the entry points, `require("./foo")` correctly resolves `foo.ts` if it's not the entry point).